client/OPUNetGameSelectWnd.cpp

```diff
@@ -469,171 +468,184 @@ void OPUNetGameSelectWnd::OnReceive(Packet &packet)
+	SetStatusText("Join Failed:  The requested game is full");
+
+	joiningGame = nullptr;
+}
+
+bool OPUNetGameSelectWnd::ReceiveJoin(Packet& packet)
+{
+	// Verify packet size
+	if (packet.header.sizeOfPayload != sizeof(JoinReply)) {
+		return false; // Discard packet
+	}
+	// Make sure we've requested to join a game
+	if (joiningGame == nullptr)
+	{
+		SetStatusText("Unexpected Join reply received");
+		return false; // Discard packet
+	}
+	// Check the session identifier
+	if (packet.tlMessage.joinReply.sessionIdentifier != joiningGame->sessionIdentifier)
+	{
+		SetStatusText("Join reply received with wrong Session ID");
+		return false; // Discard packet
+	}
+
+	return true; // Do not discard packet
+}
+
+void OPUNetGameSelectWnd::ReceiveEchoExternalAddress(Packet& packet)
+{
+	// Verify packet size
+	if (packet.header.sizeOfPayload != sizeof(EchoExternalAddress)) {
+		return;						// Discard packet
+	}
+	// Verify packet came from game server **TODO**
+
+	// Check where the information came from
+	if (packet.tlMessage.echoExternalAddress.replyPort == internalPort)
+	{
+		bReceivedInternal = true;
+	}
+	// Check if we've received a different external port than previous
+	if (ntohs(packet.tlMessage.echoExternalAddress.addr.sin_port) != externalPort)
+	{
+		if (externalPort != 0)
 		{
-			// Build new net info text string
-			std::string text = "External IP: " + FormatIP4Address(externalIp.s_addr) + ":" + std::to_string(externalPort);
-			// Check if internal address received
-			if (bReceivedInternal)
-			{
-				// Not quite true, since internal port might be random (no hosting, but possibly still open)
-				//text += " (Direct Host Capable)";
-			}
-			if (bTwoExternal)
-			{
-				text += "\nWarning: Address and Port-Dependent Mapping detected\nYou may have difficulty joining games.";
-			}
-			// Update net info text
-			SendDlgItemMessage(this->hWnd, IDC_NetInfo, WM_SETTEXT, 0, (long)text.c_str());
+			bTwoExternal = true;	// Received two distinct external ports
 		}
+	}
+	// Record external information
+	externalIp = packet.tlMessage.echoExternalAddress.addr.sin_addr;
+	if ((externalPort == 0) || (externalPort == internalPort))
+	{
+		externalPort = ntohs(packet.tlMessage.echoExternalAddress.addr.sin_port);
+	}
 
-		return;							// Packet handled
-	default:  // Silence warnings about unused enumeration value in switch
-		break;
+	// Build new net info text string
+	std::string text = "External IP: " + FormatIP4Address(externalIp.s_addr) + ":" + std::to_string(externalPort);
+	// Check if internal address received
+	if (bReceivedInternal)
+	{
+		// Not quite true, since internal port might be random (no hosting, but possibly still open)
+		//text += " (Direct Host Capable)";
 	}
+	if (bTwoExternal)
+	{
+		text += "\nWarning: Address and Port-Dependent Mapping detected\nYou may have difficulty joining games.";
+	}
+	// Update net info text
+	SendDlgItemMessage(this->hWnd, IDC_NetInfo, WM_SETTEXT, 0, reinterpret_cast<LPARAM>(text.c_str()));

```

Comment:
Besides switching to a reinterpret cast, this is also casting to LPARAM instead of long now.